### PR TITLE
add exponential backoff for retryable CSR error in nodeagent

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/9035 

to avoid traffic spike to hit CA

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -49,6 +53,12 @@ const (
 
 	// identityTemplate is the format template of identity in the CSR request.
 	identityTemplate = "spiffe://%s/ns/%s/sa/%s"
+
+	// For REST APIs between envoy->nodeagent, default value of 1s is be used.
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: "is used"

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -447,10 +457,28 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	var certChainPEM []string
+	for true {
+		certChainPEM, err = sc.fetcher.CaClient.CSRSign(ctx, csrPEM, exchangedToken, int64(sc.configOptions.SecretTTL.Seconds()))
+		if err == nil {
+			break
+		}
+
+		isTimeout := startTime.Add(time.Millisecond * envoyDefaultTimeoutInMilliSec).Before(time.Now())
+		// If non-retryable error or reach envoy timeout, fail the request by returning err
+		if !isRetryableErr(status.Code(err)) || isTimeout {
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: The logic in this not accurate in this if block as we are catching non-retryable error or timeout, however, we never log/return timeout error.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 1 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: updated.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 2 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: updated.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -447,10 +457,33 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+
+		// If non-retryable error, fail the request by returning err
+		if !isRetryableErr(status.Code(err)) {
+			log.Errorf("CSR for %q hit non-retryable error %v", resourceName, err)
+			return nil, err
+		}
+
+		// If reach envoy timeout, fail the request by returning err
+		if startTime.Add(time.Millisecond * envoyDefaultTimeoutInMilliSec).Before(time.Now()) {
+			log.Errorf("CSR retry timeout for for %q: %v", resourceName, err)
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Remove extra 'for'

**Pull request merged**